### ISSUE TITLE
Pattern-matching lambdas in erased contexts are sometimes erased, sometimes not

### ISSUE THREAD

**Issue by nad** [agda/agda#4525]
The following code is accepted:
```agda
postulate
  F : @0 Set → Set

G : @0 Set → Set
G A = F (λ { → A })
```
However, the following code is rejected:
```agda
@0 F : @0 Set → Set
F A = λ { → A }
```
I think that Agda should be consistent, and always make pattern-matching lambdas in erased contexts erased.

**Comment by andreasabel** [comment]
The comment says that extended lambdas are for now checked in non-erased context:
https://github.com/agda/agda/blob/63fe8a8e0df6141931452572b95360e129828a5c/src/full/Agda/TypeChecking/Rules/Term.hs#L662-L668
The pointer to `test/Succeed/Issue{3581}.agda` must mean:
https://github.com/agda/agda/blob/63fe8a8e0df6141931452572b95360e129828a5c/test/Succeed/Issue3581.agda#L40-L43
The difference between your examples can be explained with a _resurrection_ action on the context when going into an erased context.  I admit the story isn't round yet.

**Comment by jespercockx** [comment]
@andreasabel Any plans to work on this before the release?

**Comment by nad** [comment]
@jespercockx, @andreasabel and I discussed this. One option is to make all (current) pattern-matching lambdas non-erased, and to introduce new notation for erased pattern-matching lambdas: `λ @0 {`, `λ @0 where`.

**Comment by nad** [comment]
I'm in the process of adding erased pattern-matching lambdas. However, that does not address the fact that the following code is accepted:

```agda
postulate
  F : @0 Set → Set

G : @0 Set → Set
G A = F (λ { → A })
```

The reason why this code is accepted is that when an erased argument is checked, then Agda makes all variables in the context non-erased:

https://github.com/agda/agda/blob/b63b0ad0e4f1af7e001342bebb0c3daaf542f328/src/full/Agda/TypeChecking/Rules/Application.hs#L694-L699

Thus, when the pattern-matching lambda is checked, the variable `A` is treated as non-erased.

The following code is rejected, because Agda does not make the variables in the context non-erased:
```agda
@0 F : @0 Set → Set
F A = λ { → A }
```

I suggest that we stop changing the context, and instead keep track of whether a term is in an erased context or not. (The `TCEnv`'s `envModality` field is already supposed to keep track of this.) Any thoughts about this?

**Comment by jespercockx** [comment]
Would this change have any effect on code that does not use pattern-matching lambdas?

**Comment by nad** [comment]
I don't know. If it does, then that might point to a bug. I suggest that we run some experiments.

**Comment by nad** [comment]
When I ran a bunch of tests I only found one problem. Agda fails to solve a constraint in the following piece of code:
```agda
open import Agda.Builtin.Sigma
open import Agda.Builtin.Equality

_∘_ : {A : Set} {B : A → Set} {C : {x : A} → B x → Set} →
      (∀ {x} (y : B x) → C y) → (g : (x : A) → B x) →
      ((x : A) → C (g x))
f ∘ g = λ x → f (g x)

subst : {A : Set} {x y : A} (P : A → Set) → x ≡ y → P x → P y
subst _ refl p = p

postulate
  Proofs : {A B : Set} → (A → B) → (B → A) → Set

record _≃_ (A B : Set) : Set where
  constructor ⟨_,_⟩
  field
    to             : A → B
    is-equivalence : Σ (B → A) (Proofs to)

  from : B → A
  from = fst is-equivalence

  proofs : Proofs to from
  proofs = snd is-equivalence

  postulate
    to-from : ∀ x → to (from x) ≡ x

record _≃ᴱ_ (A B : Set) : Set where
  field
    to        : A → B
    from      : B → A
    @0 proofs : Proofs to from

@0 ≃ᴱ→≃ : {A B : Set} → A ≃ᴱ B → A ≃ B
≃ᴱ→≃ A≃B = ⟨ to , (from , proofs) ⟩
  where
  open _≃ᴱ_ A≃B

postulate
  [≃]→≃ᴱ : {A B : Set} (to : A → B) (from : B → A) →
           @0 Proofs to from → A ≃ᴱ B
  A      : Set
  P      : A → Set
  A≃A    : A ≃ A

Π-cong-≃ :
  (Q : A → Set) →
  (∀ x → P x ≃ Q (_≃_.to A≃A x)) →
  ((x : A) → P x) ≃ ((x : A) → Q x)
Π-cong-≃ Q P≃Q =
  ⟨ (λ h x → subst Q (_≃_.to-from A≃A x)
               (_≃_.to (P≃Q (_≃_.from A≃A x)) (h (_≃_.from A≃A x))))
  , ( (λ h x → _≃_.from (P≃Q x) (h (_≃_.to A≃A x)))
    , easy
    )
  ⟩
  where
  postulate
    easy : _

fails :
  {@0 _ : Set}
  (Q : A → Set) →
  (∀ x → P x ≃ᴱ Q (_≃_.to A≃A x)) →
  ((x : A) → P x) ≃ᴱ ((x : A) → Q x)
fails Q P≃Q =
  [≃]→≃ᴱ
    (λ h x → subst Q (_≃_.to-from A≃A x)
               (_≃ᴱ_.to (P≃Q (_≃_.from A≃A x)) (h (_≃_.from A≃A x))))
    _
    (_≃_.proofs (Π-cong-≃ _ (≃ᴱ→≃ ∘ P≃Q)))
```
This can be fixed by, say, replacing the last underscore with `Q`.

However, something strange is going on here: If `{@0 _ : Set}` is removed from the type signature of `fails`, then the current development version of Agda also fails to solve a (very similar) constraint. Can anyone figure out why the unused, erased argument makes a difference?

**Comment by nad** [comment]
I compared the debug printout (`-vtc:20 -vtc.meta:40`) for the code above with the code where the last instance of `@0` has been removed, and it appears as if in the former case the meta-variable `_140` is η-expanded, but not in the latter case. The meta-variable is η-expanded after the following message:
```
offending variable:  _
  of type  Set
  (after singleton test)
error during flexible occurs check, we are StronglyRigid
Pattern violation (you shouldn't see this)
oops, pattern violation for _140
```
There is no occurrence of `oops` in the debug printout for the failing code.

Some information related to the instantiation of `_140` seems to be missing:
```
Debug printing tc.meta.eta:15 failed due to exception:
  An internal error has occurred. Please report this as a bug.
  Location of the error: impossible, called at src/full/Agda/TypeChecking/Rules/LHS.hs:746:95 in Agda-2.6.2-CCl7tl6DGn2DkWiAPP1T9:Agda.TypeChecking.Rules.LHS
completed assignment of _140
```
This error message refers to an empty substitution:

https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/TypeChecking/Rules/LHS.hs#L746

I don't know if this internal error is benign or indicative of a broken invariant. Is the right context used?

https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/TypeChecking/MetaVars.hs#L579-L583
